<launch>
  <node pkg="visual_servoing" exec="visual_servoing_node.py" name="visual_servoing_node" output="screen">
	  <!--subscriber topic parameter setting/-->
    <param name="odom_topic" value="/wheel_odom" /><!--里程計 Topic（map -> base_link, 是tf相對關係, 非輪式里程計）/-->
    <param name="shelf_topic" value="/apriltag_poses" /><!--車體相機對位貨架的 Topic/-->
    <param name="pallet_topic" value="/detections" /><!--牙叉相機對位棧板 Topic/-->
    <param name="forkpose_topic" value="/forklift_pose" /><!--牙叉編碼器 Topic/-->

    <!--bodycamera parking parameter setting-->
    <param name="bodycamera_tag_offset_x"  value = "0.0" /><!--對位目標點與tag的左右偏移量(有時候相機並不會正對著tag)/-->
    <param name="bodycamera_parking_fork_init"  value = "-1.0" /><!--車體相機對位初始牙叉高度, 負值代表牙叉維持原本高度-->
    <param name="bodycamera_ChangingDirection_threshold"  value = "0.015" />  <!--正對Tag(atan2(y, x) <= threshold)/-->
    <param name="bodycamera_desired_dist_threshold"  value = "0.1" />  <!--如果正對Tag後，需要移動的距離小於threshold，就不執行move_nearby_parking_lot動作-->
    <param name="bodycamera_parking_stop"  value = "0.5" />  <!--車體相機對位停止距離/-->
    <param name="bodycamera_Changingtheta_threshold"  value = "0.025" />  <!--平行Tag(theta <= threshold)/-->
    <param name="bodycamera_decide_distance"  value = "0.025" />  <!--將相機平行Tag之後，左右y距離誤差幾公尺內可以將牙叉插進去，如果不行的話就後退重新對位/-->
    <param name="bodycamera_back_distance"  value = "0.8" />  <!--重新對位時後退的距離/-->
  
    <!--forkcamera parking setting/-->
    <param name="forkcamera_parking_fork_layer1" value = "0.0" /><!--牙叉相機對位初始牙叉高度, 為了將牙叉相機與棧板高度一致/-->
    <param name="forkcamera_parking_fork_layer2" value = "0.3" /><!--牙叉相機對位初始牙叉高度, 為了將牙叉相機與棧板高度一致/-->
    <param name="forkcamera_tag_offset_x" value = "0.0" /><!--對位目標點與tag的左右偏移量/-->
    <param name="forkcamera_ChangingDirection_threshold" value = "0.01" />  <!--正對Tag(atan2(y, x) <= threshold) threshold/-->
    <param name="forkcamera_parking_stop" value = "0.34" /><!--牙叉相機對位停止距離/-->
    <param name="forkcamera_Changingtheta_threshold" value = "0.025" />  <!--平行Tag(theta <= threshold)/-->
    <param name="forkcamera_decide_distance" value = "0.025"/>  <!--將相機平行Tag之後，左右y距離誤差幾公尺內可以將牙叉插進去，如果不行的話就後退重新對位/-->
    <param name="forkcamera_back_distance" value = "1.1" />  <!--重新對位時後退的距離/-->

    <!--Raise the pallet setting/-->
    <param name="raise_pallet_fork_init_layer1" value = "0.0" /><!--插起棧板的初始牙叉高度/-->
    <param name="raise_pallet_fork_init_layer2" value = "0.39" /><!--插起棧板的初始牙叉高度/-->
    <param name="raise_pallet_dead_reckoning_dist" value = "-0.28" /><!--直線移動到距離Tag多少公尺的地方/-->
    <param name="raise_pallet_fork_forward_distance" value = "0.655" /><!--牙叉前伸的距離/-->
    <param name="raise_pallet_raise_height_layer1" value = "0.12" /><!--牙叉叉起棧板的高度/-->
    <param name="raise_pallet_raise_height_layer2" value = "0.5" /><!--牙叉叉起棧板的高度/-->
    <param name="raise_pallet_back_distance" value = "0.9" /><!--插起棧板後，後退距離/-->
    <param name="raise_pallet_navigation_helght" value = "0.12" /><!--完成插起棧板動作後，接續下一個動作時牙叉高度/-->

    <!--Drop the pallet setting/-->
    <param name="drop_pallet_fork_init_layer1" value = "0.12" /><!--放下棧板的初始牙叉高度/-->
    <param name="drop_pallet_fork_init_layer2" value = "0.5" /><!--放下棧板的初始牙叉高度/-->
    <param name="drop_pallet_dead_reckoning_dist" value = "-0.25" /><!--使用里程計計算直走(正值)或後退(負值)距離，因為牙叉升起有可能會擋到Tag，因此改用里程計計算/-->
    <param name="drop_pallet_fork_forward_distance" value = "0.70" /><!--牙叉前伸的距離/-->
    <param name="drop_pallet_drop_height_layer1" value = "0.0" /><!--牙叉放下棧板的高度/-->
    <param name="drop_pallet_drop_height_layer2" value = "0.39" /><!--牙叉放下棧板的高度/-->
    <param name="drop_pallet_back_distance" value = "0.9" /><!--放下棧板後，後退距離/-->
    <param name="drop_pallet_navigation_helght" value = "0.0" /><!--完成放置棧板動作後，接續下一個動作時牙叉高度/-->
  </node>
</launch>
